File: BeerViewer/Models/Wrapper/ShipInfo.cs
using System.Linq;

using BeerViewer.Models.Raw;
using BeerViewer.Models.Enums;
using BeerViewer.Models.kcsapi.mst;

namespace BeerViewer.Models.Wrapper
{
	public class ShipInfo : RawDataWrapper<kcsapi_mst_ship>, IIdentifiable
	{
		private ShipType _ShipType;

		public static int[] AircraftShipTypes { get; } = new[] { 6, 7, 10, 11, 14, 16, 18, 20, 22 };

		#region Wrapping
		public int Id => base.RawData.api_id;
		public int SortId => this.RawData.api_sortno;

		public string Name => this.RawData.api_name;
		public ShipType ShipType => this._ShipType
			?? (ShipType)(this._ShipType = Master.Instance.ShipTypes[this.RawData.api_stype])
			?? null;

		public int[] Slots => this.RawData.api_maxeq;
		public ShipSpeed Speed => (ShipSpeed)this.RawData.api_soku;

		public int? NextRemodelingLevel => this.RawData.api_afterlv == 0 ? null : (int?)this.RawData.api_afterlv;

		public string Kana => this.RawData.api_yomi;

		public int HP => this.RawData.api_taik.Get(0) ?? 0;
		public int MaxArmor => this.RawData.api_souk.Get(1) ?? 0;

		public int MaxFirepower => this.RawData.api_houg.Get(1) ?? 0;
		public int MaxTorpedo => this.RawData.api_raig.Get(1) ?? 0;
		public int MaxAA => this.RawData.api_tyku.Get(1) ?? 0;

		public int MinLuck => this.RawData.api_luck.Get(0) ?? 0;
		public int MaxLuck => this.RawData.api_luck.Get(1) ?? 0;

		public int MaxFuel => this.RawData.api_fuel_max;
		public int MaxBull => this.RawData.api_bull_max;

		public int SlotCount => this.RawData.api_slot_num;
		#endregion

		public bool IsAirCraft => ShipInfo.AircraftShipTypes.Contains(this.ShipType?.Id ?? -1);

		internal ShipInfo(kcsapi_mst_ship api_data) : base(api_data) { }

		public override string ToString()
			=> $"ID = {this.Id}, Name = \"{this.Name}\", ShipType = \"{this.ShipType?.Name}\"";


		private static ShipInfo _Empty { get; } = new ShipInfo(new kcsapi_mst_ship
		{
			api_id = 0,
			api_name = "???"
		});
		public static ShipInfo Empty() => _Empty;
	}
}

File: WindowFrame/source/System/Master/Wrappers/ShipInfo.ts
import { RawDataWrapper } from "System/Base/Wrapper";
import { kcsapi_mst_ship } from "System/Interfaces/Master/kcsapi_mst_ship";
import { IIdentifiable } from "System/Base/Interfaces/IIdentifiable";
import { ShipType, ShipSpeed } from "System/Enums/ShipEnums";
import { Master } from "System/Master/Master";
import { MasterMinMax } from "System/Master/MasterMinMax";
import { ShipTypeInfo } from "./ShipTypeInfo";

export class ShipInfo extends RawDataWrapper<kcsapi_mst_ship> implements IIdentifiable {
    //#region Wrapping
    public get Id(): number { return this.raw.api_id }
    public get SortId(): number { return this.raw.api_sortno }

    public get Name(): string { return this.raw.api_name }
    public get ShipType(): ShipTypeInfo | null { return Master.Instance.ShipTypes!.get(this.raw.api_stype) || null }

    public get Slots(): number { return this.raw.api_slot_num }
    public get SlotAircraftCounts(): number[] { return this.raw.api_maxeq }
    public get Speed(): ShipSpeed { return this.raw.api_soku }

    public get NextRemodelLevel(): number | null { return this.raw.api_afterlv || null }

    public get HP(): MasterMinMax { return new MasterMinMax(this.raw.api_taik) }
    public get Armor(): MasterMinMax { return new MasterMinMax(this.raw.api_souk) }
    public get FirePower(): MasterMinMax { return new MasterMinMax(this.raw.api_houg) }
    public get Torpedo(): MasterMinMax { return new MasterMinMax(this.raw.api_raig) }
    public get AA(): MasterMinMax { return new MasterMinMax(this.raw.api_tyku) }
    public get Luck(): MasterMinMax { return new MasterMinMax(this.raw.api_luck) }

    public get Fuel(): number { return this.raw.api_fuel_max }
    public get Ammo(): number { return this.raw.api_bull_max }
    //#endregion

    public IsAircraftShipType(): boolean { return this.SlotAircraftCounts.some(x => x > 0) }

    constructor(api_data: kcsapi_mst_ship) {
        super(api_data);
    }

    public toString(): string {
        return `{"Id": ${this.Id}, "Name": "${this.Name}", "ShipType": ${this.ShipType}}`
    }

    public static readonly Empty: ShipInfo = new ShipInfo({
        api_id: 0,
        api_sortno: 0,
        api_name: "???",
        api_yomi: "???",
        api_stype: 0,
        api_afterlv: 0,
        api_aftershipid: "",
        api_taik: [0, 0],
        api_souk: [0, 0],
        api_houg: [0, 0],
        api_raig: [0, 0],
        api_tyku: [0, 0],
        api_luck: [0, 0],
        api_soku: 0,
        api_leng: 0,
        api_slot_num: 0,
        api_maxeq: [],
        api_buildtime: 0,
        api_broken: [0, 0, 0, 0],
        api_powup: [0, 0, 0, 0],
        api_backs: 0,
        api_getmes: "",
        api_afterfuel: 0,
        api_afterbull: 0,
        api_fuel_max: 0,
        api_bull_max: 0,
        api_voicef: 0
    });
}